### PR TITLE
[Aichat] Fetch student chat history and display in workspace

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds the frontend api call to fetch student chat history and follows up https://github.com/code-dot-org/code-dot-org/pull/60256 which added the backend controller action. It also displays the teacher view of student chat history in the aichat workspace.

Note that UI improvements/updates (auto-scroll to last message, ability to show/hide profane user messages, ...) will be taken care of in a follow-up PR.

Reminder that the teacher view of the student chat is hidden behind an experiment flag: 
'?enableExperiments=view_chat_history'.

## After update

Screencast video of teacher view of section with 3 students at '/s/allthethings/lessons/47/levels/1':

https://github.com/user-attachments/assets/e07f168f-f176-42f6-a7f8-9fb39b87958c

Screencast video of teacher view of section with 3 students on a sublevel at '/s/allthethings/lessons/52/levels/8/sublevel/1':

https://github.com/user-attachments/assets/5e6f750a-084d-4a11-a6e6-7194e993abc1

Note that for `chatEvent`s that have property `hideForParticipants` set to `true` (e.g., 'The user clears the chat workspace'), these events are NOT visible in the workspace. However, when a teacher is viewing the student chat history, the property `isTeacherView` is set to `true` so that these `chatEvent`s are visible in the workspace.

Teacher view of student chat history:

Screenshot 2024-08-12 at 9 27 46 AM

Screencast video of user as participant - chat events are not shown if `hideForParticipants` is set to `true` (e.g., 'The user clears the chat workspace.', 'The user loads the aichat level.').

https://github.com/user-attachments/assets/3c522a5e-83ce-4b85-8059-d16cbb5f49be

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-938)

## Testing story
Added a dashboard unit test for checking that expected chat events returned by the `student_chat_history` endpoint.
Tested locally on aichat levels in /allthethings including the aichat sublevel at '/s/allthethings/lessons/52/levels/8/sublevel/1'

## Follow-up work
UI updates to teacher view of student chat history [jira](https://codedotorg.atlassian.net/browse/LABS-968).

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/aichat/aichatApi.ts
Comment (fisher-alice) [inline apps/src/aichat/aichatApi.ts:10]: Alphabetized.

File: dashboard/test/controllers/aichat_controller_test.rb
Comment (fisher-alice) [inline dashboard/test/controllers/aichat_controller_test.rb:204]: Resolves comment from https://github.com/code-dot-org/code-dot-org/pull/60256#discussion_r1710063351

File: apps/src/aichat/aichatApi.ts
Comment (thomasoniii) [inline apps/src/aichat/aichatApi.ts]: this is fine, but could also just be:
```
if (scriptLevelid) {
  params.scriptLevelId = scriptLevelId.toString();
}
```

that way `params` could be declared as `const` on line 116 and we wouldn't be copying and creating a new object (which is super cheap and all, but still)

File: apps/src/aichat/redux/aichatRedux.ts
Comment (thomasoniii) [inline apps/src/aichat/redux/aichatRedux.ts]: I think this could be safely moved up inside the `try{}` block.

That way you could move the declaration of `studentChatHistoryApiResponse` inside the `try` block as well and make it a const.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (thomasoniii) [inline apps/src/aichat/views/ChatWorkspace.tsx]: keys on `index` values are frowned upon. Is there something in the `event` object we could key off of instead? Maybe `event.timestamp`? I'm unclear if that's unique enough. A combo of `timestamp + descriptionKey`?

**Review by thomasoniii — APPROVED** [review]
couple of tiny structural suggestions, but overall lgtm 🚀

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx]: Thanks! I think `timestamp` would be unique enough so will update.

File: apps/src/aichat/redux/aichatRedux.ts
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts]: Oh nice! Will update.

File: apps/src/aichat/aichatApi.ts
Comment (fisher-alice) [reply apps/src/aichat/aichatApi.ts]: Nice - will update.

File: apps/src/util/HttpClient.ts
Comment (sanchitmalhotra126) [inline apps/src/util/HttpClient.ts]: Any reason to not just use `fetchJson`?

File: apps/src/aichat/redux/aichatRedux.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/redux/aichatRedux.ts]: Oh does the server send back stringified JSON? Could we parse the JSON the backend before sending so we don't have to parse here? We could also then type the API return type as `ChatEvent[]` I think, right?

File: apps/src/aichat/types.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/types.ts]: How is this different from `hideForParticipants`? Do we need both?

Edit: seems like this is needed to override the check for `hideForParticipants` in `ChatItemView` if I'm reading correctly. Included a suggestion below on a potential refactor where we might not need this.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatWorkspace.tsx]: Alright, another refactoring suggestion 😅 this is coming out of a couple things I noticed:
- `StudentChatHistoryView` and `ChatWithModel` are almost identical (other than adding `isTeacherView` to the events in the former)
- Relatedly, it seems like we need `isTeacherView` to override the check for `hideForParticipants` in `ChatEventView`
- Looking at this again, I'm also wondering if the student's chat history even needs to be in Redux if it's only going to be used and updated in one place (whichever component renders the history).

So, here's a potential suggestion to hopefully make things easier to follow:
- Remove the check for `hideForParticipants` in `ChatEventView`. `ChatEventView` simply renders the given `ChatEvent`.
- Create a component like `<ChatWithModel>` that simply renders a list of `ChatEvent`s in a container. This can just replace `ChatWithModel` or even replace `ChatEventView` (i.e. turn `ChatEventView` into some `ChatEventsList` that renders a list of `ChatEvent`s in a `div` container). We can also move autoscroll logic to this container so the student chat history gets it for free.
- We can then have a separate component for Student Chat History that 1) fetches and stores (in state) the list of chat event history from the server on mount or whenever dependencies change (level ID, channel ID, user ID, etc) and 2) uses the above component to render the list of returned `ChatEvent`. No need to add a separate `isTeacherView` flag since `ChatItemView` will just render whatever is handed to it
- To display the current user's chat with the model, get the list of user events from redux and filter out anything with `hideForParticipants: true`. Use the above `<ChatEventsList>` component to render this filtered list of events.

My thinking is that this would result in less conditional logic, more reused components, and a bit of separation between the actual items in the student chat history and the items in the current chat with the model.

**Review by sanchitmalhotra126 — COMMENTED** [review]
So cool to see the history coming back from the server!

A few minor comments and refactor suggestion but the approach looks right to me.

File: apps/src/util/HttpClient.ts
Comment (fisher-alice) [reply apps/src/util/HttpClient.ts]: Yeah - I was considering using this as well, but unsure if we wanted to be able to add the authenticity token or not.

Comment (fisher-alice) [reply apps/src/util/HttpClient.ts]: I'll go ahead and use `fetchJson` - just did some reading and see that generally we don't need CSRF tokens for GET requests. Thanks!

File: apps/src/util/HttpClient.ts
Comment (fisher-alice) [inline apps/src/util/HttpClient.ts:142]: Alphabetize

File: apps/src/aichat/types.ts
Comment (fisher-alice) [reply apps/src/aichat/types.ts]: I was thinking that we need a `isTeacherView` property so that we can allow teachers to toggle between hide/view student messages that are tagged as profane. But will hold off on adding until follow-up and will see if I can implement via props vs adding another property.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx]: Thanks Sanchit!
I made the following updates based on your suggestions above:
- Removed the internal `StudentChatHistoryView` and `ChatWithModel` components.
- Added `ChatEventsList` component that is used to display the teacher view of student chat history and the visible chat events.
- Removed `isTeacherView` for now - see https://github.com/code-dot-org/code-dot-org/pull/60322#discussion_r1715695998
- Changed name of selector `selectAllMessages` to `selectAllVisibleMessages` and then removed condition to check for `hideForParticipants`
- Kept `studentChatHistory` stored in redux because of some of the complex logic (checking if sublevel) and the async call, and already checking student list for selected student name in Teacher panel. (discussed this offline).

File: apps/src/aichat/aichatApi.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/aichatApi.ts]: Why do we need to create two `URLSearchParams`? Could we just do `` `${STUDENT_CHAT_HISTORY_URL}?${new URLSearchParams(params)}` ``

File: apps/src/aichat/views/ChatEventsList.tsx
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatEventsList.tsx:13]: Not your change, but passing a function instead of a boolean here reads a little weird to me (maybe because previously it was an internal function?). Not necessary for this PR, but maybe we could change this to a `showWaitingAnimation: boolean` and have this component render the appropriate waiting image if true?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

One question on the screen recordings - it looks like the notification events in the history view have a close button. Does this mean they're dismissible?

File: apps/src/aichat/views/ChatEventsList.tsx
Comment (fisher-alice) [reply apps/src/aichat/views/ChatEventsList.tsx:13]: Oh yeah - I agree a `boolean` would be expected here. I'll update in follow-up!

File: apps/src/aichat/aichatApi.ts
Comment (fisher-alice) [reply apps/src/aichat/aichatApi.ts]: Oh yeah! Think that's just an oversight. Will update.

**Comment by fisher-alice** [comment]
> LGTM!
> 
> One question on the screen recordings - it looks like the notification events in the history view have a close button. Does this mean they're dismissible?

Great question! Interestingly, they're not dismissible. Looking into why they're not able to be dismissed, when the user clicks on the 'x' button, `removeUpdateMessage` is called and then `modelUpdateMessageInfo` is assigned `undefined` by `getUpdateMessageLocation` which takes an `id` and `state`. Don't fully understand this, but `state` is assigned a `Proxy(Object)` so that `messageList` is a `Proxy(Array)` so that no messages are found.

Do you have any ideas on why `Proxy(Object)` is found here?

FYI - in the follow-up, I'll remove the 'x' close buttons.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (fisher-alice) [inline apps/src/aichat/views/ChatWorkspace.tsx:72]: FYI - added `currentLevelId` so that student chat history is updated when level changes (after pr approval) cc: @sanchitmalhotra126

File: dashboard/app/controllers/aichat_controller.rb
Comment (fisher-alice) [inline dashboard/app/controllers/aichat_controller.rb:155]: Want the oldest event to be first, not last since student chat history lists events from oldest to most recent.

**Pull request merged**
